private/protocol/json/jsonutil/unmarshal.go

```diff
@@ -17,16 +16,10 @@ import (
 func UnmarshalJSON(v interface{}, stream io.Reader) error {
 	var out interface{}
 
-	b, err := ioutil.ReadAll(stream)
-	if err != nil {
-		return err
-	}
-
-	if len(b) == 0 {
-		return nil
-	}
-
-	if err := json.Unmarshal(b, &out); err != nil {
+	if err := json.NewDecoder(stream).Decode(&out); err != nil {
+		if err == io.EOF {
```

Comment:
Did you find a case where the `io.EOF` was being returned, and it wasn't a parsing error?  I wouldn't of expected `Decode` to return an `EOF` error if it successfully read the content.

I think the `if err == io.EOF {` block can be removed.

Reply:
This may be compatible with the below in the previous code.
```
if len(b) == 0 {
	return nil
}
```
But I cloudn't confirm where the scope is the same:thinking:

Reply:
@atsushi-ishibashi - Oof, I really don't like relying on the `io.EOF` as it requires internal knowledge on how `Decode` works. However, I think we can accept this change with one minor tweak

```go
err := json.NewDecoder(stream).Decode(&out)
if err == io.EOF {
    return nil
} else if err != nil {
    return err
}
```

It may be a good idea of not creating a new decoder every time either. We can let `UnmarshalJSON` handle the creation of the stream and pass that into `unmarshalJSON`. Let me know what you think of that!
```go
// or maybe make this a method on an object? That way decoder doesn't need to be passed
// down
func unmarshalJSON(decoder json.Decoder, v interface{}, stream io.Reader) error {
    err := decoder.Decode(stream)
    // same logic as above
}
```

Reply:
> I really don't like relying on the io.EOF as it requires internal knowledge on how Decode works.

Completely agree👍I'll modify the way of error handling following with your suggestion.

> It may be a good idea of not creating a new decoder every time either.

I don't agree with passing decoder as an argument because it will change interface. Currently `UnmarshalJSON` is package's public fuction so one possible choice is that creating decoder as package variable in `init` fuction or variable initialization. But we have to consider that there is a way to reassign `io.Reader` to `Docoder`. As far as I examined in a few minutes, it doesn't seem to exist:thinking: 
When I misunderstand your suggestion, please correct me!

Reply:
Yes, I agree, we shouldn't change the signature for `UnmarshalJSON`! What I was thinking is creating a non-exported function `unmarshalJSON` and `UnmarshalJSON` is used to just instantiate the decoder. I hope that makes sense!